gnosis/verification.k

```diff
@@ -72,22 +72,52 @@ module VERIFICATION
 
 // Reading symbolic WordStack from memory, then extracting bytes from it. General case (stack size can be != 32)
 //========================================================================
-    rule nthbyteof(chop(#asWord(W : WS) &Int MASK), I       , N       )
-      => nthbyteof(chop(#asWord(    WS) &Int MASK), I -Int 1, N -Int 1)
-      requires I >Int 0
+    //decoding nthbyteof(#asWord())
+    rule nthbyteof(#asWord(W : WS), I       , N       )
+      => nthbyteof(#asWord(    WS), I -Int 1, N -Int 1)
+      requires I >Int 0 andBool #sizeWordStack(W : WS) <=Int 32
 
-    rule nthbyteof(chop(#asWord(W : WS) &Int MASK), 0, N) => W &Int ((MASK >>Byte #sizeWordStack(WS)) &Int 255)
+    rule nthbyteof(#asWord(W : WS), 0, N) => W
+      requires #sizeWordStack(W : WS) <=Int 32
+
+    //decoding nthbyteof(#asWord() &Int MASK)
+    rule nthbyteof(#asWord(W : WS) &Int MASK, I       , N       )
+      => nthbyteof(#asWord(    WS) &Int MASK, I -Int 1, N -Int 1)
+      requires I >Int 0 andBool #sizeWordStack(W : WS) <=Int 32
+
+    rule nthbyteof(#asWord(W : WS) &Int MASK, 0, N) => W &Int (MASK >>Byte #sizeWordStack(WS))
+      requires #sizeWordStack(W : WS) <=Int 32
+
+    //decoding nthbyteof() &Int MASK
+    rule MASK &Int nthbyteof(WS, I, N) => nthbyteof(WS, I, N) &Int MASK
+      requires #isConcrete(MASK)
+
+    rule nthbyteof(_, _, _)  &Int (A => A &Int 255)
+      requires A >Int 255
```

Comment:
i cannot see what it means. can you explain?

Reply:
It trims the mask to the last byte only. Because nthbyteof has only one byte. Side condition is required to avoid rewriting to itself.